listings/advanced-concepts/simple_account/src/simple_account.cairo

```diff
@@ -0,0 +1,102 @@
+    use core::num::traits::Zero;
+    use core::ecdsa::check_ecdsa_signature;
+    use starknet::crypto::hash::starknet_keccak;
+
+    #[storage]
+    struct Storage {
+        public_key: felt252,
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, public_key: felt252) {
+        self.public_key.write(public_key);
+    }
+
+    #[abi(embed_v0)]
+    impl SRC6 of ISRC6<ContractState> {
+        fn __execute__(self: @ContractState, calls: Array<Call>) -> Array<Span<felt252>> {
+            assert(starknet::get_caller_address().is_zero(), 'Not Starknet Protocol');
+            let Call { to, selector, calldata } = calls.at(0);
+            let res = starknet::syscalls::call_contract_syscall(*to, *selector, *calldata).unwrap();
+            array![res]
+        }
+
+        fn __validate__(self: @ContractState, calls: Array<Call>) -> felt252 {
+            assert(starknet::get_caller_address().is_zero(), 'Not Starknet Protocol');
+            let tx_info = starknet::get_tx_info().unbox();
+            let tx_hash = tx_info.transaction_hash;
+            let signature = tx_info.signature;
+            if (self._is_valid_signature(tx_hash, signature)) {
+                starknet::VALIDATED
+            } else {
+                0
+            }
+        }
+
+        fn is_valid_signature(
+            self: @ContractState, hash: felt252, signature: Array<felt252>
+        ) -> felt252 {
+            if self._is_valid_signature(hash, signature.span()) {
+                starknet::VALIDATED
+            } else {
+                0
+            }
+        }
+    }
+
+    #[generate_trait]
+    impl SignatureVerificationImpl of SignatureVerification {
+        fn _is_valid_signature(
+            self: @ContractState, hash: felt252, signature: Span<felt252>
+        ) -> bool {
+            check_ecdsa_signature(
+                hash, self.public_key.read(), *signature.at(0_u32), *signature.at(1_u32)
+            )
+        }
+    }
+    
+    #[abi(embed_v0)]
+    impl SRC5 of ISRC5<ContractState> {
+        fn supports_interface(self: @ContractState, interface_id: felt252) -> bool {
+            // Calculate interface IDs for ISRC5 and ISRC6
+            let isrc5_id = starknet_keccak("supports_interface(felt252)->E((),())");
+            let isrc6_id_1 = starknet_keccak("__execute__(Array<(ContractAddress,felt252,Array<felt252>)>)->Array<(@Array<felt252>)>");
+            let isrc6_id_2 = starknet_keccak("__validate__(Array<(ContractAddress,felt252,Array<felt252>)>)->felt252");
+            let isrc6_id_3 = starknet_keccak("is_valid_signature(felt252,Array<felt252>)->felt252");
+
+            let isrc6_id = isrc6_id_1 ^ isrc6_id_2 ^ isrc6_id_3;
+
+            let supported_interfaces = array![isrc5_id, isrc6_id];
+
+            for id in supported_interfaces.iter() {
+                if *id == interface_id {
+                    return true;
+                }
+            }
+            false
+        }
+    }
+}
```

Comment:
Great! Actually, as it's not the main topic an implementation with OZ and interface ids as constant might be better. See [OZ doc](https://docs.openzeppelin.com/contracts-cairo/0.14.0/introspection#computing_the_interface_id).

But this is a great example to show SRC5, so if you are interested there's still #109 to explain SRC5 and show a similar implementation.

Reply:
thank you for the review i would make necessary changes to the implementation and i would be interested in the other issue

Reply:
Perfect, you can change this implementation with OZ SRC5 and work on the SRC5 example after.
Also, there's still tests missing here. Let me know if you want me or someone else to add theses.

Reply:
yes you can help with the test that would be great

Reply:
@julio4 please can you kindly review